File: algorithms/trees/splay-tree/typescript/splayTree.ts
class SplayNode {
    key: number;
    left: SplayNode | null = null;
    right: SplayNode | null = null;
    constructor(key: number) { this.key = key; }
}

function rightRotate(x: SplayNode): SplayNode {
    const y = x.left!;
    x.left = y.right;
    y.right = x;
    return y;
}

function leftRotate(x: SplayNode): SplayNode {
    const y = x.right!;
    x.right = y.left;
    y.left = x;
    return y;
}

function splayOp(root: SplayNode | null, key: number): SplayNode | null {
    if (!root || root.key === key) return root;
    if (key < root.key) {
        if (!root.left) return root;
        if (key < root.left.key) {
            root.left.left = splayOp(root.left.left, key);
            root = rightRotate(root);
        } else if (key > root.left.key) {
            root.left.right = splayOp(root.left.right, key);
            if (root.left.right) root.left = leftRotate(root.left);
        }
        return root.left ? rightRotate(root) : root;
    } else {
        if (!root.right) return root;
        if (key > root.right.key) {
            root.right.right = splayOp(root.right.right, key);
            root = leftRotate(root);
        } else if (key < root.right.key) {
            root.right.left = splayOp(root.right.left, key);
            if (root.right.left) root.right = rightRotate(root.right);
        }
        return root.right ? leftRotate(root) : root;
    }
}

function insertNode(root: SplayNode | null, key: number): SplayNode {
    if (!root) return new SplayNode(key);
    root = splayOp(root, key)!;
    if (root.key === key) return root;
    const node = new SplayNode(key);
    if (key < root.key) {
        node.right = root;
        node.left = root.left;
        root.left = null;
    } else {
        node.left = root;
        node.right = root.right;
        root.right = null;
    }
    return node;
}

function inorderCollect(node: SplayNode | null, result: number[]): void {
    if (!node) return;
    inorderCollect(node.left, result);
    result.push(node.key);
    inorderCollect(node.right, result);
}

export function splayTree(arr: number[]): number[] {
    let root: SplayNode | null = null;
    for (const val of arr) root = insertNode(root, val);
    const result: number[] = [];
    inorderCollect(root, result);
    return result;
}

File: algorithms/trees/splay-tree/rust/splay_tree.rs
type Link = Option<Box<SNode>>;

struct SNode {
    key: i32,
    left: Link,
    right: Link,
}

impl SNode {
    fn new(key: i32) -> Self {
        SNode { key, left: None, right: None }
    }
}

fn right_rotate(mut x: Box<SNode>) -> Box<SNode> {
    let mut y = x.left.take().unwrap();
    x.left = y.right.take();
    y.right = Some(x);
    y
}

fn left_rotate(mut x: Box<SNode>) -> Box<SNode> {
    let mut y = x.right.take().unwrap();
    x.right = y.left.take();
    y.left = Some(x);
    y
}

fn splay_op(root: Link, key: i32) -> Link {
    let mut root = match root {
        None => return None,
        Some(r) => r,
    };
    if root.key == key {
        return Some(root);
    }
    if key < root.key {
        if root.left.is_none() {
            return Some(root);
        }
        let mut left = root.left.take().unwrap();
        if key < left.key {
            left.left = splay_op(left.left.take(), key);
            root.left = Some(left);
            root = right_rotate(root);
        } else if key > left.key {
            left.right = splay_op(left.right.take(), key);
            if left.right.is_some() {
                let rotated = left_rotate(left);
                root.left = Some(rotated);
            } else {
                root.left = Some(left);
            }
        } else {
            root.left = Some(left);
        }
        if root.left.is_some() {
            Some(right_rotate(root))
        } else {
            Some(root)
        }
    } else {
        if root.right.is_none() {
            return Some(root);
        }
        let mut right = root.right.take().unwrap();
        if key > right.key {
            right.right = splay_op(right.right.take(), key);
            root.right = Some(right);
            root = left_rotate(root);
        } else if key < right.key {
            right.left = splay_op(right.left.take(), key);
            if right.left.is_some() {
                let rotated = right_rotate(right);
                root.right = Some(rotated);
            } else {
                root.right = Some(right);
            }
        } else {
            root.right = Some(right);
        }
        if root.right.is_some() {
            Some(left_rotate(root))
        } else {
            Some(root)
        }
    }
}

fn insert_node(root: Link, key: i32) -> Box<SNode> {
    match root {
        None => Box::new(SNode::new(key)),
        Some(r) => {
            let mut r = splay_op(Some(r), key).unwrap();
            if r.key == key {
                return r;
            }
            let mut node = Box::new(SNode::new(key));
            if key < r.key {
                node.left = r.left.take();
                node.right = Some(r);
            } else {
                node.right = r.right.take();
                node.left = Some(r);
            }
            node
        }
    }
}

fn inorder(node: &Link, result: &mut Vec<i32>) {
    if let Some(ref n) = node {
        inorder(&n.left, result);
        result.push(n.key);
        inorder(&n.right, result);
    }
}

pub fn splay_tree(arr: &[i32]) -> Vec<i32> {
    let mut root: Link = None;
    for &val in arr {
        root = Some(insert_node(root, val));
    }
    let mut result = Vec::new();
    inorder(&root, &mut result);
    result
}
